### ISSUE TITLE
IterableDataset.filter() checkpoint/resume silently drops the rest of the in-flight Arrow table

### ISSUE THREAD

**Issue by ebarkhordar** [huggingface/datasets#8359]
### Describe the bug

Resuming a streaming `IterableDataset.filter()` from a `state_dict()` checkpoint silently drops rows: the remainder of the in-flight Arrow table is skipped. For a single-shard dataset this loses up to a full row group per checkpoint, and for a single-table shard it can drop every unconsumed row.

`filter()` wraps the Arrow source in `FormattedExamplesIterable` but, unlike `map()`, does not first rebatch it to one row. `FormattedExamplesIterable` takes the Arrow branch and pulls a whole table from the source, advancing the shard position to the end of that table while emitting examples one at a time. The default non-batched path never records a rewind checkpoint, so `state_dict()` saves the shard position at the table end and resume skips every row of that table that had not yet been emitted.

Batched `filter()`, `map()`, `skip`, `take`, `select_columns`, `rename_column` and formatting all resume without loss under the same harness; only non-batched `filter()` on an Arrow-backed source is affected. This is the standard `load_dataset(..., streaming=True).filter(...)` + `StatefulDataLoader` training-resume path.

### Steps to reproduce the bug

```python
from datasets import Dataset

n = 10000
ds = Dataset.from_dict({"a": list(range(n))}).to_iterable_dataset(num_shards=1).filter(lambda e: True)

it = iter(ds)
seen = [next(it)["a"] for _ in range(1500)]  # consume 1500
state = ds.state_dict()

ds2 = Dataset.from_dict({"a": list(range(n))}).to_iterable_dataset(num_shards=1).filter(lambda e: True)
ds2.load_state_dict(state)
rest = [e["a"] for e in ds2]

print(len(seen) + len(rest))     # 9500, not 10000
print(sorted(set(range(n)) - set(seen + rest))[:5])  # [1500, 1501, 1502, 1503, 1504]
```

Consuming 1500 drops rows 1500..1999 (the rest of the second 1000-row table). Consuming 10 drops 990 rows.

### Expected behavior

`seen + rest` should equal every row exactly once, as it already does for batched `filter()` and for `map()`. No row should be lost across a checkpoint/resume.

### Environment info

- `datasets` version: 5.0.1.dev0 (main, commit 521a590)
- Python: 3.11.15
- pyarrow: 25.0.0
- Reproduced in a clean `python:3.11-slim` container.

**Comment by ebarkhordar** [comment]
#self-assign
